### PR TITLE
Support tuple specialisation

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

**Comment by dwijnand** [comment]
Despite trying not to to start with, I ended up needing to allow for the parsing of specialised tuple (and product) classes, so that I could reach for the class and their constructor.  Without that it ends up in a weird state where it gets removed as soon as you touch it ("implode-on-touch") due to, I expect, the deletion and markAbsent that using `NoEmbedded` leads to.

But now I realised I need to replace `x._1()` with `x._1$mcI$sp()` too, but the receiver is Tuple2 and now I can't find that method either - because I'm guessing it's in the bytecode and not the pickle...  Getting to be quite the drag...

**Comment by sjrd** [comment]
Could this be disabled when the `-scalajs` flag is on? In Scala.js, specialization does nothing but generate more code.

**Comment by dwijnand** [comment]
Yep: `override def isEnabled(using Context): Boolean = !ctx.settings.scalajs.value`

**Comment by dwijnand** [comment]
@sjrd any chance you could help me land this?  I worked on it because it was desired for 3.2, and I believe we want to cut the RC1 for that next Tuesday (and Martin is traveling atm).

**Comment by dwijnand** [comment]
@sjrd 

> What is it blocked on? / How would you like me to help?

It needs a review to be merged.

File: tests/pos/i11114.scala [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class Foo {
+  val x: (Int, Int) = (1, 1)
+  val y: Int = x._1
+}
```
Comment (sjrd) [inline tests/pos/i11114.scala:4]: Perhaps create a bytecode test in `DottyBytecodeTests.scala` with that snippet, to ensure that `x._1` is indeed compiled as `x._1$mpI$sp` or whatever it's called.

File: compiler/src/dotty/tools/dotc/transform/SpecializeTuples.scala [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      cpy.Apply(tree)(Select(New(defn.SpecialisedTuple(fun.symbol.owner.companionClass, targs.map(_.tpe)).typeRef), nme.CONSTRUCTOR), args).withType(tree.tpe)
+    case Apply(TypeApply(fun: NameTree, targs), args)
+        if fun.name == nme.CONSTRUCTOR && fun.symbol.exists && defn.isSpecializableTuple(fun.symbol.owner, targs.map(_.tpe)) =>
+      cpy.Apply(tree)(Select(New(defn.SpecialisedTuple(fun.symbol.owner, targs.map(_.tpe)).typeRef), nme.CONSTRUCTOR), args).withType(tree.tpe)
+    case _ => tree
+  end transformApply
+
+  override def transformSelect(tree: Select)(using Context): Tree = tree match
+    case Select(qual, nme._1) if isAppliedSpecializableTuple(qual.tpe.widen)
+    => Select(qual, nme._1.specializedName(qual.tpe.widen.argInfos.slice(0, 1)))
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/transform/SpecializeTuples.scala:36]: Hum, I've never seen this formatting. Shouldn't it be
```suggestion
    case Select(qual, nme._1) if isAppliedSpecializableTuple(qual.tpe.widen) =>
      Select(qual, nme._1.specializedName(qual.tpe.widen.argInfos.slice(0, 1)))
```
?

File: compiler/src/dotty/tools/dotc/transform/SpecializeTuples.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  import tpd.*
+
+  override def phaseName: String                 = SpecializeTuples.name
+  override def description: String               = SpecializeTuples.description
+  override def isEnabled(using Context): Boolean = !ctx.settings.scalajs.value
+
+  override def transformApply(tree: Apply)(using Context): Tree = tree match
+    case Apply(TypeApply(fun: NameTree, targs), args)
+        if fun.name == nme.apply && fun.symbol.exists && defn.isSpecializableTuple(fun.symbol.owner.companionClass, targs.map(_.tpe)) =>
+      cpy.Apply(tree)(Select(New(defn.SpecialisedTuple(fun.symbol.owner.companionClass, targs.map(_.tpe)).typeRef), nme.CONSTRUCTOR), args).withType(tree.tpe)
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/transform/SpecializeTuples.scala:27]: Shouldn't we test that the `qualifier` of `fun` can safely be elided here?

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 4]
```diff
@@ -1610,6 +1629,13 @@ class Definitions {
   @tu lazy val Function2SpecializedReturnClasses: PerRun[collection.Set[Symbol]] =
     new PerRun(Function2SpecializedReturnTypes.map(_.symbol))
 
+  def isSpecializableTuple(base: Symbol, args: List[Type])(using Context): Boolean =
+    args.length <= 2 && base.isClass && TupleSpecializedClasses.exists(base.asClass.derivesFrom) && args.match
+      case List(x)    => Tuple1SpecializedParamClasses().contains(x.typeSymbol)
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:1634]: Don't you have to use `x.widenDealias` here?

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:1634]: Let me see if I can break it without.  What made you think of that?

File: compiler/src/dotty/tools/dotc/transform/SpecializeTuples.scala (hunk 3 above)
Comment (sjrd) [reply compiler/src/dotty/tools/dotc/transform/SpecializeTuples.scala:27]: Although contrived, it's possible to construct something like
```scala
val t = { println("foo"); Tuple2 }.apply(1, 2)
```
in which case I think your code will throw away the `println("foo")`. It's possible that this is already broken in `main`, though. I think some stuff with `{ println("foo"); Int }.box(5)` is broken in Scala 2 as well. If that's the case, perhaps at least add a comment?

File: compiler/src/dotty/tools/dotc/core/Definitions.scala (hunk 4 above)
Comment (sjrd) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:1634]: Usually it's suspicious to take `.typeSymbol` on an arbitrary `Type` before erasure. Singleton types, or intersection types that involve `& Int`, for example, may cause trouble here.

File: compiler/src/dotty/tools/dotc/transform/SpecializeTuples.scala (hunk 3 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/transform/SpecializeTuples.scala:27]: And is there already something that lights up with that prefix?  Is it the "pure" (or is it "idempotent"?) stuff I've noticed but not looked at?

**Review by sjrd — APPROVED** [review]
LGTM. Sorry for the delay.

**Review by odersky — CHANGES_REQUESTED** [review]
Looks good otherwise.

**Review by odersky — APPROVED** [review]
Great to get this in!

**Comment by SethTisue** [comment]
Curious where/how it was decided to stop at arity 2?

**Comment by dwijnand** [comment]
It was decided in the standard library.  Tuple1 has 3 specialised classes: Double, Int and Long.  Tuple2 has 25! (Int, Long, Double, Char, Boolean) x (Int, Long, Double, Char, Boolean)...

**Comment by Adam-Vandervorst** [comment]
You're talking about the maximum number of specialised classes right @dwijnand?
The examples use known types (e.g. `foo: (Int, Int)`) for which I assume only one class is generated.
Even in many generic instances, the tuple types are not independent (e.g. `pos: (N, N, N)`) and only need 5 specializations.
In some other contexts, I can see it being useful to deduce or manually specify the cases (e.g. when N: Integral).

**Pull request merged**